tools/ci-operator-prowgen/main.go

```diff
@@ -0,0 +1,318 @@
+// Example
+// -------
+// - agent: kubernetes
+//   always_run: true
+//   branches:
+//   - <branch>
+//   context: ci/prow/<test.Name>
+//   decorate: true
+//   name: pull-ci-<org>-<repo>-<test.Name>
+//   rerun_command: /test <test.Name>
+//   skip_cloning: true
+//   spec: <generatePodSpec>
+//   trigger: ((?m)^/test( all| <test.Name>),?(\\s+|$))
+func generatePresubmitForTest(test testDescription, org, repo, branch string) *prowconfig.Presubmit {
+	presubmit := prowconfig.Presubmit{
+		Agent:        "kubernetes",
+		AlwaysRun:    true,
+		Brancher:     prowconfig.Brancher{Branches: []string{branch}},
+		Context:      fmt.Sprintf("ci/prow/%s", test.Name),
+		Name:         fmt.Sprintf("pull-ci-%s-%s-%s", org, repo, test.Name),
+		RerunCommand: fmt.Sprintf("/test %s", test.Name),
+		Spec:         generatePodSpec(org, repo, branch, test.Target),
+		Trigger:      fmt.Sprintf("((?m)^/test( all| %s),?(\\\\s+|$))", test.Name),
+		UtilityConfig: prowconfig.UtilityConfig{
+			DecorationConfig: &prowkube.DecorationConfig{SkipCloning: true},
+			Decorate:         true,
+		},
+	}
+	return &presubmit
+}
+
+// Generate a Presubmit job for the given parameters
+// Example
+// - agent: kubernetes
+//   decorate: true
+//   name: branch-ci-<org>-<repo>-<test.Name>
+//   skip_cloning: true
+//   spec: <generatePodSpec>
+func generatePostsubmitForTest(test testDescription, org, repo, branch string, additionalArgs ...string) *prowconfig.Postsubmit {
+	postsubmit := prowconfig.Postsubmit{
+		Agent: "kubernetes",
+		Name:  fmt.Sprintf("branch-ci-%s-%s-%s", org, repo, test.Name),
+		Spec:  generatePodSpec(org, repo, branch, test.Target, additionalArgs...),
+		UtilityConfig: prowconfig.UtilityConfig{
+			DecorationConfig: &prowkube.DecorationConfig{SkipCloning: true},
+			Decorate:         true,
+		},
+	}
+	return &postsubmit
+}
+
+// Given a ci-operator configuration file and basic information about what
+// should be tested, generate a following JobConfig:
+//
+// - one presubmit and one postsubmit for each test defined in config file
+// - if the config file has non-empty `images` section, generate an additinal
+//   presubmit and postsubmit that has `--target=[images]`. This postsubmit
+//   will additionally pass `--promote` to ci-operator and it will have
+//   `artifacts: images` label
+func generateJobs(
+	configSpec *cioperatorapi.ReleaseBuildConfiguration,
+	org, repo, branch string,
+) *prowconfig.JobConfig {
+
+	orgrepo := fmt.Sprintf("%s/%s", org, repo)
+	presubmits := map[string][]prowconfig.Presubmit{}
+	postsubmits := map[string][]prowconfig.Postsubmit{}
+
+	imagesTest := false
+
+	for _, element := range configSpec.Tests {
+		// Check if config file has "images" test defined to avoid name clash
+		// (we generate the additional `--target=[images]` jobs name with `images`
+		// as an identifier, but a user can have `images` test defined in his
+		// config file which would result in a clash)
+		if element.As == "images" {
```

Comment:
Which config files have an explicit images target like this?

Reply:
(maybe we can just remove that)

Reply:
None. But that word is not that uncommon and once someone calls his test like this, the generator would *silently* not generate something that it would otherwise, which is a frustrating error.

Reply:
I think we should add validation to the `ci-operator` that disallows such a test target name